src/main/java/org/terasology/launcher/repositories/JenkinsClient.java

```diff
@@ -19,23 +21,25 @@
 import java.util.Optional;
 import java.util.Properties;
 
-public class JenkinsClient {
+class JenkinsClient {
 
     private static final Logger logger = LoggerFactory.getLogger(JenkinsClient.class);
 
     private static final String ARTIFACT = "artifact/";
     private final Gson gson;
 
-    public JenkinsClient(Gson gson) {
+    JenkinsClient(Gson gson) {
         this.gson = gson;
     }
 
-    public Jenkins.ApiResult request(URL url) {
+    Jenkins.ApiResult request(URL url) {
         Preconditions.checkNotNull(url);
         try (BufferedReader reader = new BufferedReader(new InputStreamReader(url.openStream()))) {
             return gson.fromJson(reader, Jenkins.ApiResult.class);
+        } catch (JsonSyntaxException | JsonIOException e) {
+            logger.warn("Failed to read JSON from '{}'", url.toExternalForm(), e);
         } catch (IOException e) {
-            e.printStackTrace();
+            logger.warn("Failed to read from URL: {}", url.toExternalForm(), e);
         }
         return null;
```

Comment:
This is the sort of thing where I'd be happy to let errors _throw exceptions_ instead of returning nullable. Sure, it could help to catch gson's exception here and wrap it in something defined in our API, so the caller doesn't have to worry if it's a gson.JsonSyntaxException vs some other implementation, but having to do null-checking isn't _better_ for our caller than having to catch an exception.

Not required for this PR though.

Reply:
I'd like to go more in the direction of [monadic error handling](https://medium.com/@huund/monadic-error-handling-1e2ce66e3810), so instead of catch-and-log-to-nullable I'd like to wrap the error/message in a better error type and use that in further processing. This may be combined with new features such as [pattern matching on instanceof](https://docs.oracle.com/en/java/javase/15/language/pattern-matching-instanceof-operator.html).

Not sure how to proceed on that front, though. You can find a very simple [Result class](https://github.com/bgerstle/result-java), or go for more full-fledged libraries such as https://www.vavr.io/ or [jOOλ](https://github.com/jOOQ/jOOL). 🤓 

How much of that would we get "for free" when using Kotlin? ... or maybe exceptions aren't that bad at all, and I should befriend them a little more? 🧐

Reply:
Kotlin doesn't have Result built-in. If you want to go the full functional programming route with Kotlin, I think the thing to use is [Λrrow](https://arrow-kt.io/).

I don't think exceptions are such a bad thing. The feeling I get is that most people don't think Java's _checked exceptions_ are a feature worth keeping — e.g. Kotlin omits that idea entirely — but in general, throwing an exception when the method can't continue as planned, allowing that exception to bubble up through the stack, and having dedicated syntax for error-handling blocks (`try`/`catch`) seems okay to me.

(Clearly some language designers feel otherwise, as neither Go nor Rust have exceptions.)

Anyway, it sounds like neither of us are really fans of having `null` represent an error case. I haven't been won over by the idea of using `Result` in Java yet; it feels like it adds more noise to the happy-path (`.getOrElse`, etc) than try/catch does, but if we had a fully functional framework for use with JavaFX I could go along with that.

As this is a network operation, I also would want this to be an asynchronous function and would use something like ListenableFuture for that. But I hear programming languages have support for coroutines these days so maybe that's the cool thing to do?